### PR TITLE
fpm registry initialise

### PR DESCRIPTION
**Pull request by henilp105**
This PR initialises the repository and also sets a docker environment with a mongodb and flask app.

I would like to request all the Community members to review and add suggestions on this intial implementation.

Thanks and Regards,
Henil

CC @awvwgk @fortran-lang/admins @fortran-lang/fpm

### CODE REVIEW

**Review by awvwgk — APPROVED** [review]

**Comment by henilp105** [comment]
This PR sets up all the user login crud functions and the docker container with flask , nginx and mongodb.

Thanks and Regards,
Henil

CC @awvwgk @fortran-lang/admins @fortran-lang/fpm @arteevraina @minhqdao @perazz

**Comment by henilp105** [comment]
I have also added search , upload package , user main home page in this PR.

**Comment by henilp105** [comment]
I have also added search , upload package , user main home page in this PR.

**Review by everythingfunctional — APPROVED** [review]
I'm not well versed in constructing http requests by hand, so I've mostly just reviewed the code. I think it looks reasonable based on the limited amount of web-dev I've done. You'll want to make sure and use different templates for the login and signup pages pretty quick though.

**Comment by henilp105** [comment]
Thanks for the review and Sure sir, I have patched the login and signup templates. @everythingfunctional

**Comment by milancurcic** [comment]
Thanks, @henilp105. A few suggestions for the route layout:

1. Auth-related endpoints (`/signup`, `/login`, `/logout`, `/reset-password`, `/forgotpassword`): Consider putting them under `/auth/*` root, i.e. `/auth/signup`, `/auth/login`, etc. It will make a more readable API.
2. `/forgotpassword` -> `/forgot-password` for consistency with `/reset-password`.
3. `/<username>/*` -> `/users/<username>/*` . More RESTful and avoid clashes with top-level routes (if any).
4. Likewise for packages, i.e. consider `GET:/search/<query>` to be `GET:/packages` where different query parameters allow you to search for a subset of packages. `GET:/packages` without query parameters should return all packages in the registry.
5. Instead of `POST:/<username>/upload` for uploading a package, `POST:/packages` is more restful and seems more natural to me.

In summary, something like this:

* `/auth`:
  - `GET,POST:/auth/login`
  - `GET,POST:/auth/signup`
  -  `POST:/auth/logout`
  -  `POST:/auth/forgot-password`
  -  `POST:/auth/reset-password`
* `/users`:
   - `GET:/users`
   - `GET:/users/<username>`
* `/packages`:
  - `GET:/packages`: Get all packages
  - `GET:/packages?query=<query-string>`: Filter packages by query string
  - `GET:/packages/<package-id>`: Get a specific package
  - `POST:/packages`: Create a new package
  - `PUT:/packages`: Update/edit an existing package

Some of these endpoints may be out of scope for this PR.

**Comment by henilp105** [comment]
Thanks alot for the review, I will surely patch suggestions. @milancurcic

**Review by arteevraina — APPROVED** [review]
Changes look good to me. And I think we should discuss @milancurcic 's suggestions and update the existing #5 with the new updates. So, that we can refer it when ever we implement an API.

**Comment by henilp105** [comment]
@arteevraina I will be integrating the suggestions soon.

**Comment by henilp105** [comment]
@milancurcic I could not get the use of 
1. `GET:/users` , we already have `GET:/users/<username>` , we would not want to show the name/details of all the users in a registry, it would be both computationally intensive and privacy concerning, have I misunderstood the use/significance of this route. 
2. `GET:/packages` , we would not want to show all the packages in the registry at once, it would be computationally intensive, may be we could rate limit it to some n packages at once.I had also been considering we should have some labels like trending packages, new releases section on `/` route.

**Comment by everythingfunctional** [comment]
1. I think it's pretty common to be able to search/find users and see some information on their profile page. We could have a setting for users to choose not to be included in search results or have any information public on their profile page. In any case I think a `GET` request for the `users` route does make sense.
2. It might be interesting to have a default ordering for this route besides alphabetical. Something like trending is a good idea.

For both routes, it definitely makes sense to paginate them, as you are correct, you don't want to be querying and sending the whole set of either. The behavior and queries for the pagination are pretty common, so should be easy to find in tutorials and docs. It usually makes sense to include query parameters in the URL like `/users?page=3&sorted_by=joined&search_string=somebody`

File: flask/user.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                {"name": {"$regex": query}},
+                {"tags": {"$in": [query]}},
+                {"description": {"$regex": query}},
+            ]
+        }
+    )
+
+    return jsonify([package for package in packages])
+
+@app.route("/packages", methods=["POST"])
```
Comment (arteevraina) [inline flask/user.py:43]: How about creating a new file like ``package.py`` for the routes related to the package.

Comment (henilp105) [reply flask/user.py:43]: Sure, I had created for auth, mongo, maybe I missed that.

Comment (arteevraina) [reply flask/user.py]: Cool ! I am also rebasing my PR on top of this one in order to keep it updated.

File: flask/packages.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                {"tags": {"$in": [query]}},
+                {"description": {"$regex": query}},
+            ]
+        }
+    )
+
+    return jsonify([package for package in packages])
+
+@app.route("/packages", methods=["POST"])
+def upload(username):
```
Comment (arteevraina) [inline flask/packages.py:24]: ``username`` here should be degined in the route as well like ``/packages/<username>``

Comment (henilp105) [reply flask/packages.py:24]: we had a discussion with milan regarding this so we wanted to keep the upload to be independent of the username.

File: flask/user.py [hunk 3]
```diff
@@ -47,14 +56,46 @@ def profile(username):
 @swag_from("documentation/delete_user.yaml", methods=["POST"])
 def delete_user():
     uuid = request.form.get("uuid")
+    password = request.form.get("password")
+
     if not uuid:
-        return jsonify({"message": "User not found", "code": 401}), 401
+        return jsonify({"message": "User not found", "code": 404}), 404
```
Comment (arteevraina) [inline flask/user.py:62]: Shouldn't this be Unauthorized ?

File: flask/user.py [hunk 4]
```diff
@@ -47,14 +56,46 @@ def profile(username):
...
     else:
         user = db.users.find_one({"uuid": uuid})
 
     if not user:
         return "Invalid email or password", 401
```
Comment (arteevraina) [inline flask/user.py:67]: And this user not found ?

File: flask/user.py [hunk 5]
```diff
@@ -47,14 +56,46 @@ def profile(username):
...
+    else:
+        return jsonify({"message": "Invalid email or password", "code": 401}), 401
+
+
+@app.route("/users/account", methods=["POST"])
+@swag_from("documentation/account.yaml", methods=["POST"])
+def account():
+    uuid = request.form.get("uuid")
+    if not uuid:
+        return jsonify({"message": "User not found", "code": 401}), 401
```
Comment (arteevraina) [inline flask/user.py:86]: Unauthorized ?

File: flask/user.py [hunk 6]
```diff
@@ -47,14 +56,46 @@ def profile(username):
...
+    else:
+        user = db.users.find_one({"uuid": uuid})
 
-    db.users.delete_one({"uuid": uuid})
+    if not user:
+        return "Invalid email or password", 401
```
Comment (arteevraina) [inline flask/user.py:91]: user not found ?

File: flask/auth.py [hunk 7]
```diff
@@ -129,10 +129,18 @@ def logout():
 @swag_from("documentation/reset_password.yaml", methods=["POST"])
 def reset_password():
     password = request.form.get("password")
+    oldpassword = request.form.get("oldpassword")
     uuid = request.form.get("uuid")
```
Comment (arteevraina) [inline flask/auth.py:133]: Can you check if uuid is not None ?

File: flask/user.py (hunk 3 above)
Comment (henilp105) [reply flask/user.py:62]: sure.

File: registry/src/pages/account.js [hunk 8]
```diff
@@ -1,13 +1,253 @@
...
+  useEffect(() => {
+    if (username === null) {
+      navigate("/");
+    } else if (packages === null) {
+      dispatch(fetchPackages(username));
+
+      let formData = new FormData();
+      formData.append("uuid", uuid);
+
+      axios({
```
Comment (arteevraina) [inline registry/src/pages/account.js:45]: This should be inside accountActions.js file. The API request.

File: flask/user.py (hunk 5 above)
Comment (henilp105) [reply flask/user.py:86]: sure.

File: registry/src/pages/account.js [hunk 9]
```diff
@@ -1,13 +1,253 @@
...
+  });
+
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+    let formData = new FormData();
+    formData.append("oldpassword", oldpassword);
+    formData.append("password", password);
+    formData.append("uuid", uuid);
+
+    axios({
```
Comment (arteevraina) [inline registry/src/pages/account.js:70]: This also should be inside `accountActions.js` file. You should create an `accountReducer.js` file as well to emit the states accordingly and use `useSelector` hook to get the states.

File: flask/user.py (hunk 6 above)
Comment (henilp105) [reply flask/user.py:91]: sure.

File: registry/src/pages/account.js [hunk 10]
```diff
@@ -1,13 +1,253 @@
...
+        setError(error.response.data.message);
+      });
+  };
+  const handleDelete = async (e) => {
+    e.preventDefault();
+    let formData = new FormData();
+    formData.append("password", password);
+    formData.append("uuid", uuid);
+
+    axios({
```
Comment (arteevraina) [inline registry/src/pages/account.js:93]: Same change requested as above.

File: flask/auth.py (hunk 7 above)
Comment (henilp105) [reply flask/auth.py:133]: we have the mongodb to handle , if uuid is none then the query would be none.

File: registry/src/pages/account.js (hunk 8 above)
Comment (henilp105) [reply registry/src/pages/account.js:45]: sure.

File: registry/src/pages/account.js (hunk 9 above)
Comment (henilp105) [reply registry/src/pages/account.js:70]: sure.

File: flask/user.py (hunk 4 above)
Comment (henilp105) [reply flask/user.py:67]: sure.

File: registry/src/pages/user.js [hunk 11]
```diff
@@ -12,48 +15,98 @@ const UserPage = () => {
   const [email, setEmail] = React.useState("");
```
Comment (arteevraina) [inline registry/src/pages/user.js:14]: I think we have most of the states listed here inside `userReducer.js` ?

File: registry/src/pages/user.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+const UserPage = () => {
+  const { user } = useParams();
+  const navigate = useNavigate();
+  const [email, setEmail] = React.useState("");
+  const [dateJoined, setDateJoined] = React.useState("");
+  const [projects, setProjects] = React.useState([]);
+  const [isLoading, setIsLoading] = React.useState(true);
+
+  const url = `${process.env.REACT_APP_REGISTRY_API_URL}/users/${user}`;
```
Comment (arteevraina) [inline registry/src/pages/user.js:19]: Can you create a separate action for this inside `userActions.js` ? 

And looking at the code, this is called at first render you can use `useEffect` to call the action and pass on `isLoading` in the dependencies array.

File: registry/src/pages/user.js (hunk 11 above)
Comment (henilp105) [reply registry/src/pages/user.js:14]: @arteevraina this page is intended for the not logged in user to see the user page of owner/manager of any package(s), (like in PyPI [userpage](https://pypi.org/user/flaskcinema/))  , so we can't use the states inside `userReducer.js` ( as they are of the logged in user). 

React UI Route for this would be: `/users/:user`

File: registry/src/pages/user.js (hunk 12 above)
Comment (henilp105) [reply registry/src/pages/user.js:19]: sure.

File: registry/src/pages/user.js (hunk 11 above)
Comment (arteevraina) [reply registry/src/pages/user.js:14]: I think `userReducer.js` is not yet there in the code. You might have to create it and list these new states inside that. I think you are referring to `authReducer.js` which actually has the states of the "logged in user"

Comment (henilp105) [reply registry/src/pages/user.js:14]: sure.

File: registry/src/pages/register.js [hunk 13]
```diff
@@ -29,6 +29,7 @@ const Register = () => {
     if (isAuthenticated) {
       setCookie("uuid", uuid);
       navigate("/manage/projects");
+      window.location.reload();
```
Comment (arteevraina) [inline registry/src/pages/register.js:33]: Is it really required ? 
I think it might not be the good idea of using this as it is basically reloading the whole app that might cause optimisation issues in future. 

Why do you want to reload the page before navigating ? @henilp105

File: registry/src/components/packageItem.js [hunk 14]
```diff
@@ -45,7 +45,14 @@ const PackageItem = ({ packageEntity }) => {
         </Col>
         <Col md={4}>
           <div>
-            <h5 id="list-item-package-name">{packageEntity.name}</h5>
+            <h5 id="list-item-package-name">
+              <a
+                href={`/packages/${packageEntity.namespace}/${packageEntity.name}`}
```
Comment (arteevraina) [inline registry/src/components/packageItem.js:50]: Can you make use of ``Link`` from ``react-router`` instead of ``anchor`` tag ? 

https://stackoverflow.com/questions/43087007/react-link-vs-a-tag-and-arrow-function

File: flask/packages.py [hunk 15]
```diff
@@ -0,0 +1,611 @@
...
+            author = db.users.find_one({"_id": i["author"]})
+            i["namespace"] = namespace["namespace"]
+            i["author"] = author["username"]
+            search_packages.append(i)
+        return jsonify({"code": 200, "packages": search_packages, "total_pages": total_pages}), 200
+    else:
+        return jsonify({"status": "error", "message": "packages not found", "code": 404}), 404
+
+
+@app.route("/namespace/<namespace>", methods=["GET"])
```
Comment (arteevraina) [inline flask/packages.py:92]: How about shifting this handler function inside ``namespaces.py`` ?

File: registry/src/store/reducers/uploadReducer.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
+import {
+  UPLOAD_PACKAGE,
+  UPLOAD_PACKAGE_SUCCESS,
+  UPLOAD_PACKAGE_ERROR,
+} from "../actions/uploadActions";
+
+const initialState = {
```
Comment (arteevraina) [inline registry/src/store/reducers/uploadReducer.js:7]: There is data field here but don't see it's usage inside the reducer. There is no message field here. I think to avoid the confusion it can renamed as ``error`` if it's the error message and ``success`` if it's the success message. 

But, if you have a better idea, you should implement it.

File: registry/src/pages/upload.js [hunk 17]
```diff
@@ -1,11 +1,18 @@
...
+import Spinner from "react-bootstrap/Spinner";
 import "./upload.css";
 
 const PackageForm = () => {
   const [cookies, setCookie] = useCookies(["uuid"]);
-  const responseDiv = document.getElementById("error");
+  const dispatch = useDispatch();
+  const isLoading = useSelector((state) => state.upload.isLoading);
+  const message = useSelector((state) => state.upload.message);
+  document.getElementById("error").innerHTML = message;
```
Comment (arteevraina) [inline registry/src/pages/upload.js]: We should never do dom manipulation. Instead do conditional rendering using states like we are doing on login/signup page.

File: registry/src/store/actions/uploadActions.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+export const UPLOAD_PACKAGE_SUCCESS = "UPLOAD_PACKAGE_SUCCESS";
+export const UPLOAD_PACKAGE_ERROR = "UPLOAD_PACKAGE_ERROR";
+
+export const uploadPackage = (data) => {
+  return (dispatch) => {
+    dispatch({ type: UPLOAD_PACKAGE });
+    const url = `${process.env.REACT_APP_REGISTRY_API_URL}/packages`;
+    const form = new FormData(data);
+
+    fetch(url, {
```
Comment (arteevraina) [inline registry/src/store/actions/uploadActions.js]: You can use ``axois`` just in case to be consistent.

File: flask/auth.py [hunk 19]
```diff
@@ -28,16 +28,12 @@ def generate_uuid():
...
-        user = db.users.find_one({"email": email, "password": hashed_password})
-        uuid = generate_uuid() if user['loggedCount']==0 else user['uuid']
-    else:
-        user = db.users.find_one({"uuid": uuid})
+    email = request.form.get("email")
+    password = request.form.get("password")
+    password += salt
+    hashed_password = hashlib.sha256(password.encode()).hexdigest()
+    user = db.users.find_one({"email": email, "password": hashed_password})
+    uuid = generate_uuid() if user['loggedCount']==0 else user['uuid']
```
Comment (arteevraina) [inline flask/auth.py:36]: Backend will crash when new user directly tries to login and this statement user["loggedCount"] will result in None error. Because it is trying to access metadata from NoneType.

File: registry/src/store/actions/uploadActions.js (hunk 18 above)
Comment (arteevraina) [reply registry/src/store/actions/uploadActions.js]: Ok, I see you fixes that in later commits.

File: registry/src/store/reducers/uploadReducer.js (hunk 16 above)
Comment (arteevraina) [reply registry/src/store/reducers/uploadReducer.js:7]: This is also fixed. 👍

File: flask/auth.py (hunk 19 above)
Comment (henilp105) [reply flask/auth.py:36]: It would not crash as  I have already added https://github.com/henilp105/registry/blob/main/flask/auth.py#L131 the increment for a new user to increment `loggedCount`.

File: registry/src/pages/register.js (hunk 13 above)
Comment (henilp105) [reply registry/src/pages/register.js:33]: @arteevraina This reload is to handle the state management issues , it runs only after the navigating , it is important to reload the state , we can avoid this but it may later cause bug in refreshing the state variables ( its sort of a safety patch to avoid un authenticated access).

File: registry/src/components/packageItem.js (hunk 14 above)
Comment (henilp105) [reply registry/src/components/packageItem.js:50]: sure.

File: flask/auth.py (hunk 19 above)
Comment (arteevraina) [reply flask/auth.py:36]: You can reproduce it by entering either wrong email or password in the ``/auth/login`` API.

File: registry/src/pages/admin.js [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+  MDBModalContent,
+  MDBModalHeader,
+  MDBModalTitle,
+  MDBModalBody,
+  MDBModalFooter,
+  MDBIcon,
+} from "mdb-react-ui-kit";
+
+const AdminSection = () => {
+  const [deprecateModal, setdeprecateModal] = useState(false);
```
Comment (arteevraina) [inline registry/src/pages/admin.js:30]: These states should be stored inside reducers & actions can be written to perform any state change. Since, there are too many state variables it will become hard to maintain it here.

File: registry/src/pages/admin.js [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+//   const [isAdmin, setIsAdmin] = 
+
+  useEffect(() => {
+   
+    //   dispatch(resetErrorMessage());
+  
+  }, [isAuthenticated]);
+
+
+  const handleDeprecatePackage = () => {
```
Comment (arteevraina) [inline registry/src/pages/admin.js:131]: This and the below handler can be inside the actions.

File: registry/src/store/actions/adminActions.js [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+            statuscode: error.response.data.code,
+            message: error.response.data.message,
+        },
+    });
+  }
+};
+
+
+export const deleteUser = (username,uuid) => async (dispatch) => {
+  // Make an api call to request reset password
```
Comment (arteevraina) [inline registry/src/store/actions/adminActions.js:67]: Correct this comment.

File: registry/src/store/actions/adminActions.js [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+            statuscode: error.response.data.code,
+            message: error.response.data.message,
+        },
+    });
+  }
+};
+
+
+export const deleteNamespace = (namespace,uuid) => async (dispatch) => {
+  // Make an api call to request reset password
```
Comment (arteevraina) [inline registry/src/store/actions/adminActions.js:114]: Comment.

File: flask/user.py [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+    uuid = request.form.get("uuid")
+    if not uuid:
+        return jsonify({"message": "Unauthorized", "code": 401}), 401
+    else:
+        user = db.users.find_one({"uuid": uuid})
+
+    if not user:
+        return jsonify({"message": "User not found", "code": 404}), 404
+    
+    if "admin" not in user["roles"]:
```
Comment (arteevraina) [inline flask/user.py:135]: How are we differentiating if this a sudo admin or namespace admin ?

Comment (henilp105) [reply flask/user.py:135]: only the users who are `sudoadmin` would only have the `admin` in roles, in the namespaces we store the `namespace admins` in namespace document ( not in the user document).

File: registry/src/store/actions/adminActions.js (hunk 23 above)
Comment (henilp105) [reply registry/src/store/actions/adminActions.js:114]: sure.

File: registry/src/pages/admin.js (hunk 21 above)
Comment (henilp105) [reply registry/src/pages/admin.js:131]: sure.

File: tests/test_login.py [hunk 25]
```diff
@@ -0,0 +1,27 @@
+from base_case import BaseTestClass
+
+class TestLogin(BaseTestClass):
+
+    def test_successfull_login(self):
+        email = "arteevraina@gmail.com"
```
Comment (arteevraina) [inline tests/test_login.py:6]: For each test. It's better to create the user using signup API first. Because we never know which test will run first. So, for example: it can be the case the `test_successful_signup` test does not runs first and `test_successful_login` tests runs first. Then, it might actually fail the test because the user would not be there in the database.

Also, can you fix the spelling of ``successfull`` to ``successful`` in this and other instances as well.

https://dev.to/paurakhsharma/flask-rest-api-part-6-testing-rest-apis-4lla

File: flask/packages.py [hunk 26]
```diff
@@ -20,6 +21,22 @@
...
+
+    Returns:
+    bool: True if the string is a valid SPDX license identifier, False otherwise.
+    """
+    try:
+        licensing = get_spdx_licensing()
+        licensing.parse(license_str, validate=True)
+        return True
+    except:
+        return False
```
Comment (minhqdao) [inline flask/packages.py:39]: Nice, could you add some tests to make sure that the integration keeps working in the future?

Comment (arteevraina) [reply flask/packages.py:39]: Yes, I will update the existing tests for upload API because the API has been changed and tests are failing.

File: flask/backend.Dockerfile [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+COPY requirements.txt /src
+RUN --mount=type=cache,target=/root/.cache/pip \
+    pip3 install -r requirements.txt
+
+COPY . .
+
+CMD ["python3", "server.py"]
+
+
+# Dev environment
```
Comment (arteevraina) [inline flask/backend.Dockerfile:15]: Is this required @henilp105  ? If not, we can consider removing this and make this PR clean so that we can merge it in next few days.

Comment (henilp105) [reply flask/backend.Dockerfile:15]: Sure.

**Review by arteevraina — APPROVED** [review]
Merging this.

**Review by arteevraina — APPROVED** [review]
Merging this.
